Ember-sidenotes V1 (#1)
concordnow/ember-sidenotes: README.md

```hbs
<SidenotesWrapper @items={{this.items}} as |Sidenote item|>
  <Sidenote @id={{item.id}} @offsetY={{item.y}} @data={{item.data}} as |s|>
    <p>
      {{item.data.content}}
    </p>
    {{#if s.isActive}}
      <button type='button' {{on 'click' s.unselect}}>Unselect</button>
    {{else}}
      <button type='button' {{on 'click' s.select}}>Select</button>
    {{/if}}
  </Sidenote>
</SidenotesWrapper>
```

There are two ways of using this addon, controlled and non-controlled mode.

## Controlled & Non-controlled modes

### Controlled mode

The controlled mode makes it possible to managed which side note is selected with an argument named `selectedSidenoteId`. This has to match the argument `id` passed to a component `<Sidenote @id={{...}} />`.

As two actions `select` and `unselect` are exposed in each sidenote component, calling them will raise an action on the wrapper (respectively `onSelect` and `onUnselect`) with all the data necessary to be aware of which sidenote has been selected (`{ element, id, offsetY, data }`).

Then, this is your job to mutate your argument `selectedSidenoteId` to make the new sidenote selected.

> _MyComponent.js_

Reviewer: ```suggestion
_MyComponent.hbs_
```
Small typo ^^

Author: https://github.com/concordnow/ember-sidenotes/pull/1/commits/c23221e71e1f669c60feccbb8e13d3786e17313a

```hbs
<SidenotesWrapper
  @items={{this.items}}
  @onSelect={{this.onSelect}}
  @onUnselect={{this.onUnselect}}
  @selectedSidenoteId={{this.selectedSidenoteId}}
  as |Sidenote item|
>
  <Sidenote @id={{item.id}} @offsetY={{item.y}} @data={{item.data}} as |s|>
    {{#if s.isActive}}
      <button type='button' {{on 'click' s.unselect}}>Unselect</button>
    {{else}}
      <button type='button' {{on 'click' s.select}}>Select</button>
    {{/if}}
  </Sidenote>
</SidenotesWrapper>
```

_MyComponent.js_

```js
export default class MyComponent extends Component {
  @tracked items = [
    { id: 1, offsetY: 104, data: null },
    // ...
  ];

  @action
  onSelect({ id }) {
    this.selectedSidenoteId = id;
  }

  @action
  onUnselect() {
    this.selectedSidenoteId = null;
  }
}
```

### Non-controlled mode

> The non-controlled mode is the way to use this addon in a simpler way. The selected sidenote will be handle internally by the components.

Reviewer 2: ```suggestion
The non-controlled mode is the way to use this addon in a simpler way. The selected sidenote will be handled internally by the components.
```

Author: https://github.com/concordnow/ember-sidenotes/pull/1/commits/c0cacafe3d5f0df65da6c008eae418af7271d347

Both `select` and `unselect` actions will raised the matching action on the wrapper (respectively `onSelect` and `onUnselect`) but only for notify purpose.

> You can also specify a `defaultSidenoteId` argument to the wrapper to specify the sidenote selected on component load. If this is not specified or the one specified does not match any sidenote, this will default to the first sidenote.

Reviewer 2: ```suggestion
You can also specify a `defaultSidenoteId` argument to the wrapper to choose the sidenote selected on component load. If it's not specified or the one specified does not match any sidenote, this will default to the first sidenote.
```

_MyComponent.hbs_

```hbs
<SidenotesWrapper
  @defaultSidenoteId={{6}}
  @items={{this.items}}
  as |Sidenote item|
>
  <Sidenote @id={{item.id}} @offsetY={{item.y}} @data={{item.data}} as |s|>
    {{#if s.isActive}}
      <button type='button' {{on 'click' s.unselect}}>Unselect</button>
    {{else}}
      <button type='button' {{on 'click' s.select}}>Select</button>
    {{/if}}
  </Sidenote>
</SidenotesWrapper>
```

_MyComponent.js_

```js
export default class MyComponent extends Component {
  @tracked items = [
    { id: 1, offsetY: 104, data: null },
    // ...
    { id: 6, offsetY: 495, data: null },
  ];
}
```

## Components details

> ### SidenoteWrapper

Reviewer 2: ```suggestion
### SidenotesWrapper
```

| defaultSidenoteId | Number | null    | Non-controlled | Default selected sidenote id on wrapper load                                                                       |
| gutter            | Number | 0       | Both           | Vertical spacing added between every sidenotes                                                                     |
| items             | Array  | -       | Both           | List of items corresponding to wanted sidenotes                                                                    |
| onSelect          | Action | -       | Both           | Action raised when using `select` exposed action on `Sidenote` component                                           |
| onSidenotesMoved  | Action | -       | Both           | Action raised after sidenotes finished moving (called after load, selection, unselection, or sidenote size change) |
| onUnselect        | Action | -       | Both           | Action raised when using `unselect` exposed action on `Sidenote` component                                         |

This component expose two variables.

- [0] is a `Sidenote` component instance

> - [1] is the current item to process (as `SidenoteWrapper` iterate over `@items`)

Reviewer 2: ```suggestion
- [1] is the current item to process (as `SidenotesWrapper` iterate over `@items`)
```

### Sidenote

| Argument | Type   | Default | Description                                                                                         |
| -------- | ------ | ------- | --------------------------------------------------------------------------------------------------- |
| id       | Number | -       | Sidenote id used as the unique identifier of this sidenote (should be unique)                       |
| offsetY  | Number | -       | Preferred vertical position of the sidenote (the sidenote will position at this offset if possible) |
| data     | Any    | -       | Custom data (this will be passed to `@onSelect` and `@onUnselect` of the wrapper)                   |

This component expose one variable which has 3 properties

- `isActive` is a boolean value that reflects if current sidenote is the one selected

> - `select` is an action to use to select current sidenote (will behave differently depending on the mode used)
> - `unselect` is an action to use to unselect current sidenote (will behave differently depending on the mode used)

Reviewer 2: ```suggestion
- `onSelect` is an action to use to select current sidenote (will behave differently depending on the mode used)
- `onUnselect` is an action to use to unselect current sidenote (will behave differently depending on the mode used)
```

Author: No, Sidenote component expose 2 actions select & unselect to trigger SidenotesWrapper onSelect & onUnselect listeners
